catris.py

```diff
@@ -111,75 +122,108 @@ def get_coords(self) -> set[tuple[int, int]]:
+        return (
+            (-self.direction_y * x - self.direction_x * y),
+            (self.direction_x * x - self.direction_y * y),
+        )
+
+
+class Game:
+    landed_blocks: dict[tuple[int, int], int | None]
 
-class GameState:
     def __init__(self) -> None:
         self.reset()
 
     def reset(self) -> None:
         self.start_time = time.monotonic_ns()
         self.players: list[Player] = []
-        self._landed_blocks: list[list[int | None]] = [[] for y in range(HEIGHT)]
         self.score = 0
 
+    def is_valid(self) -> bool:
+        seen = {
+            point for point, color in self.landed_blocks.items() if color is not None
+        }
+
+        for block in self._get_moving_blocks():
+            coords = block.get_coords()
+            if coords & seen:
+                return False
+            seen.update(coords)
+
+        return True
+
     def game_is_over(self) -> bool:
         return bool(self.players) and not any(
             isinstance(p.moving_block_or_wait_counter, MovingBlock)
             for p in self.players
         )
 
-    def end_waiting(self, player: Player, client_currently_connected: bool) -> None:
-        assert player.moving_block_or_wait_counter == 0
-        if not client_currently_connected:
-            player.moving_block_or_wait_counter = None
-            return
-
-        index = self.players.index(player)
-        x_min = WIDTH_PER_PLAYER * index
-        x_max = x_min + WIDTH_PER_PLAYER
-        for row in self._landed_blocks:
-            row[x_min:x_max] = [None] * WIDTH_PER_PLAYER
-        player.moving_block_or_wait_counter = MovingBlock(index)
-
-    def get_width(self) -> int:
-        return WIDTH_PER_PLAYER * len(self.players)
-
     def _get_moving_blocks(self) -> list[MovingBlock]:
         result = []
         for player in self.players:
             if isinstance(player.moving_block_or_wait_counter, MovingBlock):
                 result.append(player.moving_block_or_wait_counter)
         return result
 
-    def is_valid(self) -> bool:
-        seen = set()
-
-        for y, row in enumerate(self._landed_blocks):
-            for x, color in enumerate(row):
-                if color is not None:
-                    seen.add((x, y))
+    def end_waiting(self, player: Player, client_currently_connected: bool) -> None:
+        assert player.moving_block_or_wait_counter == 0
+        if not client_currently_connected:
+            player.moving_block_or_wait_counter = None
+            return
+        self.wipe_playing_area(player)
 
-        for block in self._get_moving_blocks():
-            coords = block.get_coords()
-            if coords & seen or not all(
-                x in range(self.get_width()) and y < HEIGHT for x, y in coords
-            ):
-                return False
-            seen.update(coords)
-        return True
+    @abstractmethod
+    def wipe_playing_area(self, player: Player) -> None:
```

Comment:
I don't like this method. Instead of wiping it, it should probably return coordinates.